### PR TITLE
browserify-p4: playlist*, xhr, and resolve-url

### PR DESCRIPTION
**Pull request by BrandonOCasey**
updated stub.js to keep unit tests working
updated build/watch scripts
ripped resolve-url out of videojs-contrib-hls for now

### CODE REVIEW

File: src/playlist-loader.js [hunk 1]
```diff
@@ -5,376 +5,385 @@
...
-      if (!srcUrl) {
-        throw new Error('A non-empty playlist URL is required');
+
+      result.playlists[i] = mergeOptions(playlist, media);
+      result.playlists[media.uri] = result.playlists[i];
+
+      // if the update could overlap existing segment information,
+      // merge the two lists
+      if (playlist.segments) {
+        result.playlists[i].segments = updateSegments(
```
Comment (dmlap) [inline src/playlist-loader.js]: linebreak on parens

File: src/playlist-loader.js [hunk 2]
```diff
@@ -5,376 +5,385 @@
...
+    this.srcUrl = srcUrl;
+    this.withCredentials = withCredentials;
 
-        // any in-flight request is now finished
-        request = null;
-        
-        if (startingState) {
-          loader.state = startingState;
-        }
+    this.mediaUpdateTimeout = null;
```
Comment (dmlap) [inline src/playlist-loader.js]: All of these fields should be suffixed with `_` to indicate they're private.

File: src/playlist-loader.js [hunk 3]
```diff
@@ -5,376 +5,384 @@
...
       }
+      changed = true;
+    }
+  }
+  return changed ? result : null;
+};
 
-      playlistRequestError = function(xhr, url, startingState) {
-        loader.setBandwidth(request || xhr);
+export default class PlaylistLoader extends Stream {
```
Comment (dmlap) [inline src/playlist-loader.js]: Looks like there are a couple refactorings going on simultaneously in this class. Can this be split up into something like:

1. es6 stuff
2. (optional) reorder stuff

File: src/playlist.js [hunk 4]
```diff
@@ -1,198 +1,198 @@
...
+                         playlist.mediaSequence +
+                         Math.max(0, playlist.segments.length - 3));
+  return createTimeRange(start, end);
+};
+
+// exports
+export default {
+  duration,
+  seekable
+};
```
Comment (dmlap) [inline src/playlist.js]: All looks good in this file.

File: src/resolve-url.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * Constructs a new URI by interpreting a path relative to another
+ * URI.
+ * @param basePath {string} a relative or absolute URI
+ * @param path {string} a path part to combine with the base
+ * @return {string} a URI that is equivalent to composing `base`
+ * with `path`
+ * @see http://stackoverflow.com/questions/470832/getting-an-absolute-url-from-a-relative-one-ie6-issue
+ */
+/* eslint-enable max-len */
+const resolveUrl = function(basePath, path) {
```
Comment (dmlap) [inline src/resolve-url.js]: An upcoming PR will remove the implementation of this function elsewhere?

File: src/xhr.js [hunk 6]
```diff
@@ -1,48 +1,51 @@
...
-          (request && (request.response || request.responseText)));
-      }
+    // videojs.xhr no longer considers status codes outside of 200 and 0
+    // (for file uris) to be errors, but the old XHR did, so emulate that
+    // behavior. Status 206 may be used in response to byterange requests.
+    if (!error &&
+        response.statusCode !== 200 &&
+        response.statusCode !== 206 &&
+        response.statusCode !== 0) {
+      error = new Error(
```
Comment (dmlap) [inline src/xhr.js]: linebreak on parens

**Comment by dmlap** [comment]
In `playlist-loader.test.js`: pulling respond() out into its own function makes this test file different from others that use sinon and only saves 10 characters or so. Doesn't seem worth it to me.

**Comment by dmlap** [comment]
[playlist-loader.test.js#650](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R650): accidental find-replace in the error message?

**Comment by dmlap** [comment]
[playlist-loader.test.js#696](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R696): find-replace again?

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R736): linebreak on parens

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R774): linebreak on parens

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R804): fine-replace this.requests again

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R822): find-replace

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R917): linebreak parens

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R946): existing issue but duplicated assertion

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R1018): pre-existing duplicate again

**Comment by dmlap** [comment]
[playlist-loader.test](https://github.com/videojs/videojs-contrib-hls/pull/544/files?w=1#diff-6675fafcbc02876bcfe4036517fc7e36R1032): moving this comment makes its intent less clear

File: test/playlist.test.js [hunk 7]
```diff
@@ -1,371 +1,385 @@
...
+  duration = Playlist.duration(playlist, playlist.mediaSequence);
+  QUnit.equal(duration, 110.5, 'used segment end time');
+  duration = Playlist.duration(playlist, playlist.mediaSequence + 1);
+  QUnit.equal(duration, 120.5, 'used segment end time');
+  duration = Playlist.duration(playlist, playlist.mediaSequence + 2);
+  QUnit.equal(duration, 120.5 + 9, 'used segment end time');
+});
+
+QUnit.test(
+'looks outside the queried interval for live playlist timeline values',
```
Comment (dmlap) [inline test/playlist.test.js]: linebreak on parens

File: test/playlist.test.js [hunk 8]
```diff
@@ -1,371 +1,385 @@
...
+
+  duration = Playlist.duration(playlist, 2);
+  QUnit.equal(duration, 20.1, 'used the timeline-based interval');
+
+  duration = Playlist.duration(playlist, 3);
+  QUnit.equal(duration, 30.1, 'used the timeline-based interval');
+});
+
+QUnit.test(
+'counts the time between segments as part of the earlier segment\'s duration',
```
Comment (dmlap) [inline test/playlist.test.js]: linebreak on parens

File: test/playlist.test.js [hunk 9]
```diff
@@ -1,371 +1,385 @@
...
-        uri: '4.ts'
-      }]
-    }, 5);
-
-    equal(duration, 50.0002, 'calculated with mixed intervals');
+  QUnit.equal(seekable.length, 0, 'no seekable ranges from a master playlist');
+});
+
+QUnit.test(
+'seekable end is three target durations from the actual end of live playlists',
```
Comment (dmlap) [inline test/playlist.test.js]: linebreak on parens

File: test/playlist.test.js [hunk 10]
```diff
@@ -1,371 +1,385 @@
...
-    });
-    equal(seekable.start(0), 0, 'starts at the earliest available segment');
-    equal(seekable.end(0),
-          9 - (2 + 2 + 1),
-          'allows seeking no further than three segments from the end');
-  });
-
-})(window, window.videojs);
+  QUnit.equal(seekable.start(0), 0, 'starts at the earliest available segment');
+  QUnit.equal(
```
Comment (dmlap) [inline test/playlist.test.js]: linebreak on parens

**Comment by dmlap** [comment]
@BrandonOCasey finished reviewing.

File: src/resolve-url.js (hunk 5 above)
Comment (BrandonOCasey) [inline src/resolve-url.js]: yeah from the main videojs-contrib-hls file

File: src/playlist-loader.js (hunk 3 above)
Comment (BrandonOCasey) [inline src/playlist-loader.js]: yeah I will submit another PR for the reorder/refactor stuff

**Comment by BrandonOCasey** [comment]
@dmlap maybe I am just missing something but some of your comments are not attached to any lines in the playlist-loader.js file

**Comment by dmlap** [comment]
@BrandonOCasey that file didn't show up in the diff for me so I had to review it line by line. I tried to link to each of the instances but if that didn't work, you may have to re-review it as well.

File: src/playlist-loader.js [hunk 11]
```diff
@@ -5,560 +5,562 @@
  * M3U8 playlists.
```
Comment (dmlap) [inline src/playlist-loader.js]: This file looks good now.

**Comment by dmlap** [comment]
@BrandonOCasey two more line-break-on-parens in playlist.test.js but otherwise this looks good!

**Pull request merged**